website/blog/2020-02-24-presto-lambda.md

```diff
@@ -0,0 +1,193 @@
+
+The pioneering work of Dremel proposed an extension to SQL based on recursive relational algebra to allow querying nested records [1], and is now available in BigQuery and latest SQL:2016. The following example shows how to transform array elements with this (adapted from [BigQuery Docs](https://cloud.google.com/bigquery/docs/reference/standard-sql/arrays#creating-arrays-from-subqueries)): 
+
+```sql
+SELECT elements,
+    ARRAY(SELECT v * 2
+          FROM UNNEST(elements) AS v) AS multiplied_elements
+FROM (
+    VALUES
+        (ARRAY[1, 2]),
+        (ARRAY[1, 3, 9]),
+        (ARRAY[1, 4, 16, 64])
+) AS t(elements)
+
+    elements    | multiplied_elements
+----------------+---------------------
+ [1, 2]         | [2, 4]
+ [1, 3, 9]      | [2, 6, 18]
+ [1, 4, 16, 64] | [2, 8, 32, 128]
+(3 rows)
+```
+
+While nested relational algebra provides an elegant and unified approach to query nested data, we found it could be challenging for users to track the “unnest stack” in mind when writing the query. In our experience, users are more comfortable to apply a given function (e.g lambda) to each element in the collection. This motivates us to introduce lambda expression into Presto to help query nested data, as illustrated below:
+
+```sql
+SELECT elements, 
+transform(elements, v -> v * 2) as multiplied_elements
+FROM (
+    VALUES
+        (ARRAY[1, 2]),
+        (ARRAY[1, 3, 9]),
+        (ARRAY[1, 4, 16, 64])
+) AS t(elements)
+```
+
+In Presto, lambda expression consists of argument list and lambda body, separated by `->`:
+```sql
+x -> x + 1
+(x, y) -> x + y
+x -> regexp_like(x, 'a+')
+x -> x[1] / x[2]
+x -> IF(x > 0, x, -x)
+x -> COALESCE(x, 0)
+x -> CAST(x AS JSON)
+x -> x + TRY(1 / 0)
+```
+
+Note there is no type annotation in lambda expression. The type of lambda expression (e.g. `function(integer, integer)`) thus has to be inferred with the context of function call. As a result, standalone lambda expression is not allowed since its type cannot be determined. 
+
+<!--truncate-->
+
+## Lambda Type Inference
+
+The initial lambda support in Presto was added in [Presto#6198](https://github.com/prestodb/presto/pull/6198) with basic compilation and execution. One of the major challenges this pull request addressed is type inference for lambda, as there is no type annotation in lambda expression. Consider the following expression contains lambda:
+
+```sql
+transform(elements, v -> v * 2)
+```
+
+where `elements` is with type `array(integer)`.
+
+The difficulty is type for `v -> v * 2` cannot be resolved without context. And Presto cannot use name and argument types to resolve function `transform` as it normally does for other functions (more details about Presto expression type inference can be found in [this doc](https://github.com/prestodb/presto/wiki/Expression-Type-Inference)). 
```

Comment:
I think the people reading are going to want to know why and not follow a link to find out.

Reply:
@aweisberg : Indeed, you might note this wiki page is very new. I originally wrote these stuffs together in the blog post. Then I feel the "type inference" part is too long, so I "refactor" these part into a wiki page ;)